asm/ddot_test.go

```diff
@@ -0,0 +1,113 @@
+// Copyright ©2015 The gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package asm
+
+import (
+	"math"
+	"testing"
+)
+
+func TestDdotUnitary(t *testing.T) {
+	for i, test := range []struct {
+		xData []float64
+		yData []float64
+
+		want float64
+	}{
+		{
+			xData: []float64{2},
+			yData: []float64{-3},
+			want:  -6,
+		},
+		{
+			xData: []float64{0, 0, 1, 1, 2, -3, -4},
+			yData: []float64{0, 1, 0, 3, -4, 5, -6},
+			want:  4,
+		},
+	} {
+		x, _, _ := newGuardedVector(test.xData, 1)
+		y, _, _ := newGuardedVector(test.yData, 1)
+		got := DdotUnitary(x, y)
+
+		if math.IsNaN(got) {
+			t.Errorf("test %v: invalid memory access in DdotUnitary", i)
+			continue
+		}
+		if got != test.want {
+			t.Errorf("test %v: DdotUnitary failed. want %v, got %v", i, test.want, got)
+		}
+	}
+}
+
+func TestDdotInc(t *testing.T) {
+	for i, test := range []struct {
+		xData []float64
+		yData []float64
+
+		want    float64
+		wantRev float64 // Result when one of the vectors is reversed.
+	}{
+		{
+			xData:   []float64{2},
+			yData:   []float64{-3},
+			want:    -6,
+			wantRev: -6,
+		},
+		{
+			xData:   []float64{0, 0, 1, 1, 2, -3, -4},
+			yData:   []float64{0, 1, 0, 3, -4, 5, -6},
+			want:    4,
+			wantRev: -4,
+		},
+	} {
+		for _, incX := range []int{-7, -3, -2, -1, 1, 2, 3, 7} {
+			for _, incY := range []int{-7, -3, -2, -1, 1, 2, 3, 7} {
+				n := len(test.xData)
+				x, _, _ := newGuardedVector(test.xData, incX)
+				y, _, _ := newGuardedVector(test.yData, incY)
+
+				var ix, iy int
+				if incX < 0 {
+					ix = (-n + 1) * incX
+				}
+				if incY < 0 {
+					iy = (-n + 1) * incY
+				}
+				got := DdotInc(x, y, uintptr(n), uintptr(incX), uintptr(incY), uintptr(ix), uintptr(iy))
+
+				if math.IsNaN(got) {
+					t.Errorf("test %v: invalid memory access in DdotInc. incX = %v, incY = %v", i, incX, incY)
+					continue
+				}
+
+				want := test.want
+				if incX*incY < 0 {
+					want = test.wantRev
+				}
+				if got != want {
+					t.Errorf("test %v: DdotInc failed. incX = %v, incY = %v, want %v, got %v",
+						i, incX, incY, want, got)
+				}
+			}
+		}
+	}
+}
+
+func newGuardedVector(data []float64, inc int) (v, frontGuard, backGuard []float64) {
+	if inc < 0 {
+		inc = -inc
+	}
+	guard := 2 * inc
```

Comment:
The guard size should depend on `inc` and `1*inc` seems like too short. Should I make it even longer?